### PR TITLE
feat(Git Sync): UX improvements

### LINKED ISSUE #8020: High CPU and memory consumption
### Expected Behavior

like in https://github.com/Kong/insomnia/issues/8017

the beta is even worse now .. you closed the issue .. that i could not respond.

### Actual Behavior

I tried the beta like you asked me. 

the switch of branches is fast now. But CPU and esp. memory consumption is WAY WORSE! 

Screenshot 2024-09-27 112452
Screenshot 2024-09-27 112739
Screenshot 2024-09-27 112819

There must be a memoryleak somewhere. it took 4 GB memory in 1 1/2 min .. then i closed insomnia. 

### Reproduction Steps

i tried the latest beta

switch branch, fetch, switch branch ... 

### Is there an existing issue for this?

- [X] I have searched the [issue tracker](https://www.github.com/Kong/insomnia/issues) for this problem.

### Additional Information

_No response_

### Insomnia Version

latest beta 

### What operating system are you using?

Windows

### Operating System Version

10 and 11

### Installation method

insomnia website executable Beta!!

### Last Known Working Insomnia version

_No response_

### PR DESCRIPTION
**Pull request by gatzjames**
# Highlights:
- [x] Commit modal UX improvements
  - [x] Align with Insomnia Sync UX
  - [x] Diff viewer like in Insomnia Sync
  - [x] Stage/Unstage changes
  - [x] Undo pending changes (with prompt to not lose any data)
  - [x] Auto-close the modal once there are no more changes to commit

## Technical details:

Based on the architecture of isomorphic-git:
- So far we relied on:
  - Reading the workspace and it's descendants on the database
  - Ask isomorphic-git if there are any changes on any of these files
- With this change:
  - We tell isomorphic-git that there is always a .insomnia folder available
  - It automatically figures out if there are any changes and we don't need to do so in the database

This change improves the performance of checking for changes by an order of magnitude on my tests especially on large repos with hundreds of changes.

To improve performance further:
- [ ] Currently when we check for changes we read the file contents and try to parse the YAML to read it's name in order to display it on the commit modal
- [ ] We can have a separate function that only checks for changes using the git.statusMatrix function that will be much faster.

Closes INS-4486

Closes #8020

### CODE REVIEW

File: packages/insomnia/src/sync/git/__tests__/git-vcs.test.ts [hunk 1]
```diff
@@ -125,15 +250,15 @@ First commit!
...
-      expect(await GitVCS.status(fooTxt)).toBe('deleted');
+      await fsClient.promises.unlink(path.join(GIT_INSOMNIA_DIR, fooTxt));
+      // expect(await GitVCS.status(barTxt)).toBe('*added');
+      // expect(await GitVCS.status(fooTxt)).toBe('*deleted');
+      // await GitVCS.remove(fooTxt);
+      // expect(await GitVCS.status(barTxt)).toBe('*added');
+      // expect(await GitVCS.status(fooTxt)).toBe('deleted');
+      // await GitVCS.remove(fooTxt);
+      // expect(await GitVCS.status(barTxt)).toBe('*added');
+      // expect(await GitVCS.status(fooTxt)).toBe('deleted');
```
Comment (jackkav) [inline packages/insomnia/src/sync/git/__tests__/git-vcs.test.ts]: 🔪 😉 🇧🇷

File: packages/insomnia/src/sync/git/git-vcs.ts [hunk 2]
```diff
@@ -257,16 +272,269 @@ export class GitVCS {
...
+      },
     });
+
+    const diff = {
+      head: result[1],
+      workdir: result[2],
+      stage: result[3],
+    };
+
+    return diff;
```
Comment (jackkav) [inline packages/insomnia/src/sync/git/git-vcs.ts]: 😍

File: packages/insomnia/src/sync/git/git-vcs.ts [hunk 3]
```diff
@@ -257,16 +272,269 @@ export class GitVCS {
...
+
+        const isBlob = [headType, workdirType, stageType].includes('blob');
+
+        // For now, bail on directories unless the file is also a blob in another tree
+        if ((headType === 'tree' || headType === 'special') && !isBlob) {
+          return;
+        }
+        if (headType === 'commit') {
+          return null;
+        }
```
Comment (jackkav) [inline packages/insomnia/src/sync/git/git-vcs.ts:292]: is there a significance of returning null or undefined here and below? may be worth a comment if so.

File: packages/insomnia/src/sync/git/git-vcs.ts [hunk 4]
```diff
@@ -257,16 +272,269 @@ export class GitVCS {
...
+        const headOid = headType === 'blob' ? await head?.oid() : undefined;
+        const stageOid = stageType === 'blob' ? await stage?.oid() : undefined;
+        let workdirOid;
+        if (
+          headType !== 'blob' &&
+          workdirType === 'blob' &&
+          stageType !== 'blob'
+        ) {
+          // We don't actually NEED the sha. Any sha will do
+          // TODO: update this logic to handle N trees instead of just 3.
```
Comment (jackkav) [inline packages/insomnia/src/sync/git/git-vcs.ts:453]: 💰

File: packages/insomnia/src/sync/git/git-vcs.ts [hunk 5]
```diff
@@ -257,16 +272,269 @@ export class GitVCS {
...
+
+            stageBlob = blob;
+          } catch (e) {
+            console.log('[git] Failed to read blob', e);
+          }
+        }
+
+        const entry = [undefined, headOid, workdirOid, stageOid];
+        const result = entry.map(value => entry.indexOf(value));
+        result.shift();
```
Comment (jackkav) [inline packages/insomnia/src/sync/git/git-vcs.ts]: can this use the same expression as above, where you name them out of the array

File: packages/insomnia/src/ui/routes/git-actions.tsx [hunk 6]
```diff
@@ -315,22 +321,26 @@ export const gitChangesLoader: LoaderFunction = async ({
 
   const branch = await GitVCS.getCurrentBranch();
   try {
-    const { changes, statusNames } = await getGitChanges(GitVCS, workspace);
+    const { changes, hasUncommittedChanges } = await getGitChanges(GitVCS);
+
+    console.log('changes', changes);
```
Comment (jackkav) [inline packages/insomnia/src/ui/routes/git-actions.tsx]: oops

File: packages/insomnia/src/ui/routes/git-actions.tsx [hunk 7]
```diff
@@ -811,7 +846,70 @@ export const commitToGitRepoAction: ActionFunction = async ({
...
+  } catch (err) {
+    if (err instanceof Errors.HttpError) {
+      return {
+        errors: [`${err.message}, ${err.data.response}`],
+      };
+    }
+    const errorMessage = err instanceof Error ? err.message : 'Unknown Error';
+
+    return { errors: [errorMessage] };
+  }
```
Comment (jackkav) [inline packages/insomnia/src/ui/routes/git-actions.tsx:859]: its not clear why these special httperror cases exist, comment or a more meaningful variable name would help

**Review by jackkav — COMMENTED** [review]
nice

File: packages/insomnia/src/sync/git/git-vcs.ts (hunk 3 above)
Comment (gatzjames) [reply packages/insomnia/src/sync/git/git-vcs.ts:292]: Yes according to the map function docs null will not walk the tree anymore while undefined will try to walk it's children

**Review by jackkav — APPROVED** [review]
looks good, we should do a mob test on this before GA 10.1 though, likely there is an edge we can't catch by reviewing the code

**Pull request merged**